### PR TITLE
add namespace to jvm metric attributes

### LINKED ISSUE #61: Add namespaces to JVM metric attributes?
Replaces https://github.com/open-telemetry/opentelemetry-specification/issues/3417

There are (currently) three proposals being discussed for metric attribute namespaces in #51.

This purpose of this issue is to try to arrive at consensus for future JVM metric attribute renames based on each of these three proposals.

## Today:

* `process.runtime.jvm.memory.*`
  * `type`
  * `pool`
* `process.runtime.jvm.gc.duration`
  * `gc`
  * `action`
* `process.runtime.jvm.threads.count`
  * `daemon`
* `process.runtime.jvm.buffer`
  * `pool`

## Option A: No namespaces required

Possibly no change.

Maybe `gc` -> `name` to align with Option C.

## Option B: Explicit namespaces required

PR already exists for this: https://github.com/open-telemetry/semantic-conventions/pull/20

* `process.runtime.jvm.memory.*`
  * `type` -> `memory.type`
  * `pool` -> `memory.pool.name`
* `process.runtime.jvm.gc.duration`
  * `gc` -> `gc.name`
  * `action` -> `gc.action`
* `process.runtime.jvm.threads.count`
  * `daemon` -> `thread.daemon`
* `process.runtime.jvm.buffer`
  * `pool` -> `buffer.pool.name`

## Option C: Namespace inherited from metric namespace

* `process.runtime.jvm.memory.*`
  * `type` (implicitly `process.runtime.jvm.memory.type`)
  * `pool` (implicitly `process.runtime.jvm.memory.pool`)
* `process.runtime.jvm.gc.duration`
  * `gc` -> `name` (implicitly `process.runtime.jvm.gc.name`)
  * `action` (implicitly `process.runtime.jvm.gc.action`)
* `process.runtime.jvm.threads.count`
  * `daemon` (implicitly `process.runtime.jvm.threads.daemon`)
* `process.runtime.jvm.buffer`
  * `pool` (implicitly `process.runtime.jvm.buffer.pool`)

### PR DESCRIPTION
**Pull request by zeitlinger**
Successor of https://github.com/open-telemetry/opentelemetry-specification/pull/3464#pullrequestreview-1424803956

Fixes #61

### CODE REVIEW

**Review by mateuszrzeszutek — APPROVED** [review]

**Comment by trask** [comment]
From JVM runtime metrics stability WG, I think we had agreed on these attribute names if the outcome of #51 is to namespace metric attributes

**Comment by zeitlinger** [comment]
> Successor of [open-telemetry/opentelemetry-specification#3464 (review)](https://github.com/open-telemetry/opentelemetry-specification/pull/3464#pullrequestreview-1424803956)

Fixes https://github.com/open-telemetry/opentelemetry-specification/issues/3417

**Review by jsuereth — APPROVED** [review]

File: schemas/1.21.0
Comment (joaopgrassi) [inline schemas/1.21.0:79]: Is this correct?

File: CHANGELOG.md
Comment (joaopgrassi) [inline CHANGELOG.md:100]: Hi @zeitlinger is this PR still a WIP? Because the metrics in here are not the ones actually changed.

Comment (zeitlinger) [reply CHANGELOG.md]: I think I copy-pasted from the wrong lines - should be fixed now.

File: schemas/1.21.0
Comment (joaopgrassi) [inline schemas/1.21.0:56]: This change seems to be missing still. It's still just `type` in the rendered markdown.

Comment (joaopgrassi) [inline schemas/1.21.0:56]:
```suggestion
              type: memory.type
```
Isn't the attribute called `type` today?

File: CHANGELOG.md
Comment (joaopgrassi) [inline CHANGELOG.md:105]:
```suggestion
    - Rename `type` to `memory.type`
```
Maybe forgot to also change in the yaml file?

**Review by joaopgrassi — CHANGES_REQUESTED** [review]

File: schemas/1.21.0
Comment (zeitlinger) [reply schemas/1.21.0:56]: thanks for catching - this got lost rebasing

Comment (zeitlinger) [reply schemas/1.21.0:56]: it's `from:to` - and it's meant to be renamed to `memory.type`

File: semantic_conventions/metrics/process-runtime-jvm-metrics-experimental.yaml
Comment (zeitlinger) [reply semantic_conventions/metrics/process-runtime-jvm-metrics-experimental.yaml:36]: good idea - makes the intention clear

**Review by jack-berg — APPROVED** [review]
I posted a [comment](https://github.com/open-telemetry/semantic-conventions/issues/61#issuecomment-1607762924) expressing a preference for a difference strategy, but like this as well if folks are aligned that gc. / memory. / thread. / buffer. are generic enough concepts to not place in a JVM namespace.

File: semantic_conventions/metrics/process-runtime-jvm-metrics.yaml
Comment (joaopgrassi) [inline semantic_conventions/metrics/process-runtime-jvm-metrics.yaml:67]: I'm not sure if using the prefix for the metric itself is a good idea. Also, `gc` might be too broad. If we want to have other "gc" attributes, then they will collide with the ones here for this metric.

File: schemas/1.21.0
Comment (joaopgrassi) [reply schemas/1.21.0:79]: I think this is still wrong, no? @zeitlinger `attributes.process.runtime.jvm.buffer` is an attribute group, not a metric. This group applies to the metrics

`process.runtime.jvm.buffer.usage`, `process.runtime.jvm.buffer.limit` and `process.runtime.jvm.buffer.count`. 

Also: `apply_to_metrics`:

In the YAML definitions, we have the `id` (`metric.process.runtime.jvm.memory.usage`) but we also have the `metric_name` (`process.runtime.jvm.memory.usage`). If we are going to list them here, should the metric name or the id value be used? I think it's the metric name but I'm not sure. @tigrannajaryan maybe you know more?

Also: Do we want to list the metrics here, or just omit them? The [schema file format spec](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/schemas/file_format_v1.1.0.md#rename_attributes-transformation-2) says:

```
apply_to_metrics:
              # Optional. If it is missing the transformation is applied
              # to all metrics. If it is present the transformation is applied
              # only to the metrics with the name that is found in the sequence
              # specified below.
```

**Comment by joaopgrassi** [comment]
> I posted a https://github.com/open-telemetry/semantic-conventions/issues/61#issuecomment-1607762924 expressing a preference for a difference strategy, but like this as well if folks are aligned that gc. / memory. / thread. / buffer. are generic enough concepts to not place in a JVM namespace.

I also have some concerns with the broad `gc`, `memory`, `thread` prefixes. If later we want to document attributes for other GCs then they will collide with these. If they are broad enough, then it probably makes more sense to move them to a separate file, like done for http in[ `http-common.yaml`](https://github.com/open-telemetry/semantic-conventions/blob/main/semantic_conventions/http-common.yaml). You can see there the prefix is just `http` for the `attribute_group`.

File: semantic_conventions/metrics/process-runtime-jvm-metrics.yaml
Comment (zeitlinger) [reply semantic_conventions/metrics/process-runtime-jvm-metrics.yaml:67]: I thought this was just a tool to make the yaml file shorter, nothing that is changing the spec in any way.
If this is not true, then we may need to take this issue back to discussion.

File: schemas/1.21.0
Comment (zeitlinger) [reply schemas/1.21.0]: you're right - fixed

**Comment by zeitlinger** [comment]
@joaopgrassi fixed the obvious things

File: semantic_conventions/metrics/process-runtime-jvm-metrics.yaml
Comment (zeitlinger) [reply semantic_conventions/metrics/process-runtime-jvm-metrics.yaml:67]: that was not a clear yes or no answer - I'm a bit lost

Comment (joaopgrassi) [reply semantic_conventions/metrics/process-runtime-jvm-metrics.yaml:67]: It's not changing the spec, but it's kinda opening up for problems later I think. I expressed it in more detail [here](https://github.com/open-telemetry/semantic-conventions/pull/20#issuecomment-1610977656), mostly aligns with the concerns from @jack-berg I think. 

I guess we need to think about what the `memory` or `gc` namespaces will contain and if the attributes here are "generic" enough so that they can be used somewhere else. OR we pick a more specific namespace for them, like `jvm.gc`

Comment (zeitlinger) [reply semantic_conventions/metrics/process-runtime-jvm-metrics.yaml]: yes, that was very helpful :smile:

**Comment by zeitlinger** [comment]
@joaopgrassi we decided to qualify all metrics with "jvm." except for thread.daemon - everything should be addressed now

**Review by mateuszrzeszutek — APPROVED** [review]

File: semantic_conventions/metrics/thread.yaml
Comment (joaopgrassi) [inline semantic_conventions/metrics/thread.yaml:2]: Are "thread" attributes only applicable to metrics? 

```suggestion
  - id: thread
```

Or (to be consistent with https://github.com/open-telemetry/semantic-conventions/blob/main/semantic_conventions/http-common.yaml#L2

```suggestion
  - id: attributes.thread.common
```

Comment (zeitlinger) [reply semantic_conventions/metrics/thread.yaml:2]: we have thread name/id here: https://github.com/open-telemetry/semantic-conventions/blob/9e3e528cccb3a3c0c6b4d5a057b4256e398fd69f/semantic_conventions/trace/general.yaml#L202

does it make sense to move all 3 into `thread-common.yaml`?

Comment (joaopgrassi) [reply semantic_conventions/metrics/thread.yaml:2]: For the sake of the scope of this, I would keep as you have (decide what to call the new file/`id). Once this merges I will open an issue to discuss if the others in `trace/general.yaml` should be moved there. What do you think?

File: semantic_conventions/metrics/process-runtime-jvm-metrics.yaml
Comment (joaopgrassi) [inline semantic_conventions/metrics/process-runtime-jvm-metrics.yaml:92]: Please add the attribute here individually with `ref`. If later we add more attributes to that file, they will all show up here automatically and I think we don't want that.

**Review by joaopgrassi — APPROVED** [review]
Looks good once the last comments I had are solved 👍

File: semantic_conventions/metrics/thread.yaml
Comment (zeitlinger) [reply semantic_conventions/metrics/thread.yaml:2]: sounds good :smile:

File: semantic_conventions/metrics/process-runtime-jvm-metrics.yaml
Comment (zeitlinger) [reply semantic_conventions/metrics/process-runtime-jvm-metrics.yaml:92]: somehow not working - can you help?

```
Index: semantic_conventions/metrics/process-runtime-jvm-metrics.yaml
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/semantic_conventions/metrics/process-runtime-jvm-metrics.yaml b/semantic_conventions/metrics/process-runtime-jvm-metrics.yaml
--- a/semantic_conventions/metrics/process-runtime-jvm-metrics.yaml	(revision 8965737826547a304db34497943098e987b3c045)
+++ b/semantic_conventions/metrics/process-runtime-jvm-metrics.yaml	(date 1688120152579)
@@ -89,7 +89,8 @@
     brief: "Number of executing platform threads."
     instrument: updowncounter
     unit: "{thread}"
-    extends: metric_attributes.thread
+    attributes:
+      - ref: thread.daemon
 
   - id: metric.process.runtime.jvm.classes.loaded
     type: metric
```

Comment (joaopgrassi) [reply semantic_conventions/metrics/process-runtime-jvm-metrics.yaml]: @trask @zeitlinger  and I spoke and we added the `daemon` attribute under the `thread` group inside `[trace/general.yaml`](https://github.com/zeitlinger/semantic-conventions/blob/jvm_metric_attributes/semantic_conventions/trace/general.yaml#L202). All works, but that's still inside the `trace` folder. 

I was thinking if we should carve those attributes out and create a `thread-common.yaml` under `semantic_conventions`, as we did for [http-common](https://github.com/zeitlinger/semantic-conventions/blob/jvm_metric_attributes/semantic_conventions/http-common.yaml)? That way we can share the attributes on traces and metrics.

**Review by joaopgrassi — APPROVED** [review]

File: semantic_conventions/metrics/process-runtime-jvm-metrics.yaml
Comment (trask) [reply semantic_conventions/metrics/process-runtime-jvm-metrics.yaml]: makes sense to me, thx for opening #155

File: CHANGELOG.md
Comment (tigrannajaryan) [inline CHANGELOG.md:118]: Will we do a similar thing for other languages? Do we plan to allow every language to put at the top-level of the namespace a concept that applies only to that language?

I would want to understand what is our general strategy with metric attribute namespacing.

Comment (zeitlinger) [reply CHANGELOG.md:118]: > Do we plan to allow every language to put at the top-level of the namespace a concept that applies only to that language?

this PR tries to avoid exactly that.
The only attribute we put at the top level is one where we thought it would not be java specific `thread.daemon` (boolean).

Comment (tigrannajaryan) [reply CHANGELOG.md:118]: I am confused. This PR suggests `jvm.` as a top level namespace, which is specific to Java and other JVM languages. How does this PR avoid other languages doing the same and polluting the top-level namespace? Should golang use `go.gc.` namespace for GC metrics?

**Review by tigrannajaryan — CHANGES_REQUESTED** [review]
Blocking until the confusion about using top-level namespaces is clarified.

File: CHANGELOG.md
Comment (zeitlinger) [reply CHANGELOG.md:118]: Ignore my previous comment - I think I get it now :smile: 

Yes, this PR suggests that other languages should also be added to the root namespace like `jvm.`. 

I'm not sure what you mean by "polluting" the root namespace - are you suggesting something like `languages.jvm.memory.type`?
For background : The discussion in the Java SIG was that we felt that `jvm.memory.type` was already too long for practical use - but mandated by https://github.com/open-telemetry/semantic-conventions/issues/51.

Comment (tigrannajaryan) [reply CHANGELOG.md:118]: OK, going back to my original question. 

What is our general strategy with regards to choosing top-level namespaces? Is every language free to put anything in the top-level namespace (like Java does with 'jvm.`)? My concern is that this is going to result in highly crowded top-level that is going to be hard to manage. How do we make sure languages don't step on each other's toes? A language can choose to use a namespace and later it turns out to be also needed for another language but can mean something different.

I would like to understand what the policy is in general and I don't see it mentioned anywhere.

>are you suggesting something like languages.jvm.memory.type?

No, I am not suggesting anything. I am merely asking a question.

**Comment by zeitlinger** [comment]
> If it is then why not use it also for metric names, e.g. `jvm.memory.usage`?

- The `process` namespace is already used by other metrics, such as `process.cpu.utilization`
- `process.runtime.` is a useful distinction for metrics that are measured within a particular runtime, such as `metric.process.runtime.jvm.cpu.time`

**Comment by trask** [comment]
hey @tigrannajaryan, we discussed reusing the instrument namespace as the attribute namespace in https://github.com/open-telemetry/opentelemetry-specification/issues/3417#issuecomment-1516882428:

> Some (verbose) options for the other 3:
>
> * `type` -> `process.runtime.jvm.memory.type`
> * `gc` -> `process.runtime.jvm.gc.name`
> * `action` -> `process.runtime.jvm.gc.action`

the concern about this option was its verbosity level

**Comment by tigrannajaryan** [comment]
Update: I discussed this with @jack-berg and @jsuereth and we all believe it is important to come up with some guidelines on how to consistently name the namespaces for metrics and for attributes. I am going to keep this PR blocked temporarily until the guidelines like that are introduced.

**Comment by jsuereth** [comment]
For context - We'd like to ensure namespaces are chosen to be "reasonably unique".  As such, we don't think `process.runtime.jvm` is a great namespace, and simply using `jvm` for JVM metrics + attributes is acceptible as it's reasonably unique.

Will be opening a PR with this language shortly.

**Review by jmacd — APPROVED** [review]

File: schemas/1.22.0
Comment (jsuereth) [inline schemas/1.22.0:1]: This should be in the [schema-next file](https://github.com/open-telemetry/semantic-conventions/blob/main/schema-next.yaml) NOT in the schemas directory.

**Review by jsuereth — CHANGES_REQUESTED** [review]
Primarily, the schema file should NOT be in the `schemas` directory, we only create that one when cutting a release.

However, additionally to this attribute rename, we should be renaming the JVM metric namespace *as a whole* to be `jvm.*` based on a lot of discussions.


Particularly - We want namespacing to be "reasonably unique" and minimal.   `process.runtime` just doesn't make any sense for the JVM.  The JVM is not bound to one process, depending on which architecture it's run on.

File: schemas/1.22.0
Comment (zeitlinger) [reply schemas/1.22.0]: thanks - might make sense to add that to CONTRIBUTING.md

Comment (trask) [reply schemas/1.22.0]: #216 ❤️

**Comment by jsuereth** [comment]
While I still think we should entertain changing the metrics `runtime.process.jvm.` => `jvm.`, I think this makes good incremental progress.  Still need to address Tigran's comments on the schema file.

File: schema-next.yaml
Comment (trask) [reply schema-next.yaml]: done

File: schema-next.yaml
Comment (trask) [reply schema-next.yaml]: done

File: CHANGELOG.md
Comment (trask) [reply CHANGELOG.md]: resolving, moved this discussion to #227

**Comment by trask** [comment]
> Blocking until the confusion about using top-level namespaces is clarified.

@tigrannajaryan are you ok with unblocking this PR and moving the discussion to #227? or would you like to get resolution on #227 before merging this?

**Review by joaopgrassi — APPROVED** [review]

**Comment by joaopgrassi** [comment]
Depending in which one merges first (https://github.com/open-telemetry/semantic-conventions/pull/241) one of them will have to adapt either the metric names or the attribute names

**Review by trask — APPROVED** [review]

**Comment by trask** [comment]
> Depending which one merges first (#241) one of them will have to adapt either the metric names or the attribute names in the changelog/schema file

#241 got merged first, I have resolved conflicts in this PR

File: CHANGELOG.md
Comment (jack-berg) [inline CHANGELOG.md:36]: For the purposes of the changelog, do we want to reference the old metric names with the `process.runtime.*` prefix or the new shorter ones?

Comment (trask) [reply CHANGELOG.md]: makes sense to me to use the names that were in effect with this change (so the new shorter ones), I've made this change

**Comment by trask** [comment]
> However, for the attribute names we suggest `jvm.*` as the namespace. Why this and why not `process.runtime.jvm.*`? From what I understand it is because `jvm.*` is shorter. But is it unique enough to ensure there are no clashes? If it is then why not use it also for metric names, e.g. `jvm.memory.usage`?

@tigrannajaryan we have shortened the metric names also now from `process.runtime.jvm.*` to `jvm.*` (#241)

if this resolves your concern, can you unblock? thx

File: CHANGELOG.md
Comment (tigrannajaryan) [inline CHANGELOG.md:37]: Nit: let's use consistent way of listing changes (see line 30 for a different ordering).

**Review by tigrannajaryan — APPROVED** [review]

File: CHANGELOG.md
Comment (trask) [reply CHANGELOG.md:37]:
```suggestion
  - Rename attributes `gc` to `jvm.gc.name`, `action` to `jvm.gc.action`
```

**Pull request merged**
